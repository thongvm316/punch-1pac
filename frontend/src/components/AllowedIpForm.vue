<template>
  <div>
    <div class="form-group">
      <label class="form-label">{{ $t('company.allowedIPs.labels.ipAddress') }}</label>
      <input class="form-input" type="text" v-model.trim="$v.params.$model" :class="{ 'is-error': $v.params.$error, 'is-success': $v.params.isValid && params !== targetIp.ip_address }">
      <p class="form-input-hint text-error" v-if="$v.params.$error">
        {{ $t('company.allowedIPs.labels.ipAddress') }}
        <span v-if="!$v.params.required">{{ $t('company.allowedIPs.labels.blank') }}</span>
        <span v-else-if="!$v.params.isValid">{{ $t('company.allowedIPs.labels.invalid') }}</span>
      </p>
    </div>
    <div class="form-group">
      <button
        ref="createAllowedIpButton"
        class="btn btn-success btn-submit"
        @click="localAddIp"
        v-if="!targetIp"
        :disabled="isDisabled">{{ $t('company.allowedIPs.btn.submit') }}</button>
      <button
        ref="editAllowedIpButton"
        class="btn btn-success btn-submit"
        @click="localEditIp"
        v-if="targetIp"
        :disabled="isDisabled">{{ $t('company.allowedIPs.btn.save') }}</button>
    </div>
  </div>
</template>

<script>
import { mapActions } from 'vuex'
import handleSuccess from '../mixins/handle-success'
import allowedIpValidate from '../validations/allowed-ip-validate'

export default {
  name: 'allowed-ip-form',

  props: {
    targetIp: Object
  },

  mixins: [allowedIpValidate, handleSuccess],

  data() {
    return {
      data: {
        emitType: 'afterModify',
        message: ''
      },
      params: ''
    }
  },

  methods: {
    ...mapActions('companyAllowedIPs', ['createIP', 'updateIP']),

    localAddIp() {
      this.createIP({ ip_address: this.params }).then(response => {
        this.data.message = this.$t('messages.ip.createSuccess')
        this.handleSuccess(this.data)
      })
    },

    localEditIp() {
      this.updateIP({ id: this.targetIp.id, ip_address: this.params }).then(response => {
        this.data.message = this.$t('messages.ip.updateSuccess')
        this.handleSuccess(this.data)
      })
    }
  },

  computed: {
    isDisabled() {
      let flag = false

      if (this.targetIp) {
        flag = this.$v.params.$error || this.params === this.targetIp.ip_address
      } else {
        flag = this.$v.params.$error || this.params === ''
      }

      return flag
    }
  },

  created() {
    if (this.targetIp) this.params = this.targetIp.ip_address
  }
}
</script>
