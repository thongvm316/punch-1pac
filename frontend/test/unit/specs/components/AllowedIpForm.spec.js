import { shallowMount } from '@vue/test-utils'

import wrapperOps from '../../supports/wrapper'
import setComputed from '../../supports/set-computed'

import AllowedIpForm from '@/components/AllowedIpForm'

const localAddIp = jest.fn()
const localEditIp = jest.fn()
const handleSuccess = jest.spyOn(AllowedIpForm.methods, 'handleSuccess')

Object.assign(wrapperOps, {
  methods: {
    localAddIp,
    localEditIp,
    handleSuccess
  }
})

describe('AllowedIpForm.vue', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallowMount(AllowedIpForm, wrapperOps)
  })

  afterEach(() => { wrapper.vm.$destroy() })

  describe('when AllowedIpForm was mounted', () => {
    it('should display AllowedIpForm Component', () => {
      expect(wrapper.isVueInstance()).toBe(true)
      expect(wrapper.exists()).toBe(true)
    })
  })

  describe('create allowed ip button', () => {
    it('should display create allowed ip button', () => {
      expect(wrapper.vm.$data.params).toBe('')
      expect(wrapper.find({ ref: 'createAllowedIpButton' }).exists()).toBe(true)
      expect(wrapper.find({ ref: 'editAllowedIpButton' }).exists()).toBe(false)
    })

    it('should call localAddIp method', async () => {
      wrapper.find({ ref: 'createAllowedIpButton' }).trigger('click')
      await wrapper.vm.$nextTick()

      expect(localAddIp).toHaveBeenCalled()
    })
  })

  describe('update allowed ip button', () => {
    beforeEach(async () => {
      wrapper.setProps({ targetIp: '192.168.110.1' })
      await wrapper.vm.$nextTick()
    })

    it('should display update allowed ip button', () => {
      expect(wrapper.props().targetIp).toEqual('192.168.110.1')
      expect(wrapper.find({ ref: 'createAllowedIpButton' }).exists()).toBe(false)
      expect(wrapper.find({ ref: 'editAllowedIpButton' }).exists()).toBe(true)
    })

    it('should call localEditIp method', () => {
      wrapper.find({ ref: 'editAllowedIpButton' }).trigger('click')

      expect(localEditIp).toHaveBeenCalled()
    })
  })

  describe('when handleSuccess method', () => {
    it('should createSuccess', async () => {
      wrapper.vm.handleSuccess()
      await wrapper.vm.$nextTick()

      expect(wrapper.vm.isDisable).toEqual(false)
      expect(wrapper.emitted('afterModify')).toBeTruthy()
    })
  })

  describe('form input hint text', () => {
    it('should no display error text', () => {
      expect(wrapper.find('.form-group > .form-input-hint').exists()).toBe(false)
    })

    it('should display error text', async () => {
      setComputed(wrapper, { errors: { ip_address: ['cant be blank'] } })
      await wrapper.vm.$nextTick()

      expect(wrapper.find('.form-group:first-of-type').classes()).toContain('has-error')
      expect(wrapper.find('.form-group > .form-input-hint').exists()).toBe(true)
      expect(wrapper.find('.form-group > .form-input-hint').text()).toEqual('IP address cant be blank')
    })
  })
})
