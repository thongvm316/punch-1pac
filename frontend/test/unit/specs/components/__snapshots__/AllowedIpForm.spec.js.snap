// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AllowedIpForm.vue when AllowedIpForm was mounted should display AllowedIpForm Component 1`] = `
<div>
  <div class="form-group"><label class="form-label">IP address</label> <input type="text" class="form-input">
    <!---->
  </div>
  <div class="form-group"><button class="btn btn-success btn-submit">
      Submit
    </button>
    <!---->
  </div>
</div>
`;

exports[`AllowedIpForm.vue when update allowedIP should display update allowed ip button 1`] = `
<div>
  <div class="form-group"><label class="form-label">IP address</label> <input type="text" class="form-input">
    <!---->
  </div>
  <div class="form-group">
    <!----> <button class="btn btn-success btn-submit">
      Save
    </button></div>
</div>
`;
